addons/stock/models/stock_location.py

```diff
@@ -244,23 +244,31 @@ def _get_putaway_strategy(self, product, quantity=0, package=None, packaging=Non
         qty_by_location = defaultdict(lambda: 0)
         locations = self.child_internal_location_ids
         if locations.storage_category_id:
-            move_line_data = self.env['stock.move.line'].read_group([
-                ('product_id', '=', product.id),
-                ('location_dest_id', 'in', locations.ids),
-                ('state', 'not in', ['draft', 'done', 'cancel'])
-            ], ['location_dest_id', 'product_id', 'product_qty:array_agg', 'qty_done:array_agg', 'product_uom_id:array_agg'], ['location_dest_id'])
-            quant_data = self.env['stock.quant'].read_group([
-                ('product_id', '=', product.id),
-                ('location_id', 'in', locations.ids),
-            ], ['location_id', 'product_id', 'quantity:sum'], ['location_id'])
-
-            for values in move_line_data:
-                uoms = self.env['uom.uom'].browse(values['product_uom_id'])
-                qty_done = sum(max(ml_uom._compute_quantity(float(qty), product.uom_id), float(qty_reserved))
-                               for qty_reserved, qty, ml_uom in zip(values['product_qty'], values['qty_done'], list(uoms)))
-                qty_by_location[values['location_dest_id'][0]] = qty_done
-            for values in quant_data:
-                qty_by_location[values['location_id'][0]] += values['quantity']
+            if package and package.package_type_id:
+                data = self.env['stock.move.line'].read_group([
+                    ('result_package_id.package_type_id', '=', package_type.id),
+                    ('state', 'not in', ['draft', 'cancel']),
```

Comment:
You should still check the quant and the package on the quant. The current way you will have an issue in the time. The more the database will grow the more this query will be slow and it create performance issues

Reply:
This is for incoming (not done) packages, stock.quant.package won't have it's location before done. So if I check on quant for incoming ones, then I still need to check the sm lines link to them to figure out it's destination location.

Reply:
Then it's missing the done in domain no?

Reply:
Oops my bad, it's for both incoming and done packages. But anyway, it won't change the result we still need to search on sm lines.

Reply:
Then I would still use a mix of stock.quant and sml